app/Http/Controllers/Admin/StatisticsController.php

```diff
@@ -43,7 +43,11 @@ public function index(Request $request)
 
         $tokens = [];
         foreach ($nodes as $node) {
-            $tokens[$node->id] = $this->keyProviderService->handle($node->servers->get(0), $request->user());
+            $server = Server::where('node_id', $node->id)->first();
+            if ($server == null)
```

Comment:
Two things in here, the first is that this if check should use braces ( and you could do `is_null($server)`), but I would recommend just getting the count of servers from the database to check aganist.

There should also be a service that providers a node specific token, rather than a token for a specific server. All administrative users should theoretically have access to this page, and theres no admin specific permissions right now so you can expose the node daemon secret here without any issues.

Doing the per-server token will end up making a _large_ amount of DB calls if there are a lot of nodes. We can go back to this and adjust that once there are admin specific permissions in place.